### PR TITLE
add support for QUIC 38

### LINKED ISSUE #529: add support for QUIC 38
see https://codereview.chromium.org/2649953002/

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #529.

This PR contains two commits that are strictly speaking not necessary for supporting QUIC 38, but that I'd like to merge anyway:

1. It makes the `h2quic` server tests independent of the configured QUIC versions. It was always annoying to edit these tests when adding a new QUIC version.
2. A change to the `ackhandler`, which would have been necessary for eliminating the STOP_WAITING frames, as it was planned with the NSTP option. However, this isn't coming in QUIC 38 (apparently).

### CODE REVIEW

File: ackhandler/received_packet_handler.go [hunk 1]
```diff
@@ -39,31 +39,31 @@ func (h *receivedPacketHandler) ReceivedPacket(packetNumber protocol.PacketNumbe
...
-		}
-	}
-
 	if packetNumber > h.largestObserved {
 		h.largestObserved = packetNumber
 		h.largestObservedReceivedTime = time.Now()
 	}
 
+	if packetNumber <= h.lowerLimit {
+		return nil
```
Comment (lucas-clemente) [inline ackhandler/received_packet_handler.go]: This means we don't call maybeQueueAck anymore, a behavior change from before. I think that should be fine, but we won't count packets towards the 20-packet number.

File: ackhandler/received_packet_handler.go [hunk 2]
```diff
@@ -39,31 +39,31 @@ func (h *receivedPacketHandler) ReceivedPacket(packetNumber protocol.PacketNumbe
...
+	// TODO: this check is not needed anymore once we drop support for STOP_WAITINGs
+	if p == 0 {
+		return
 	}
-
-	h.ignorePacketsBelow = f.LeastUnacked - 1
-
-	h.packetHistory.DeleteBelow(f.LeastUnacked)
-	return nil
+	h.lowerLimit = p - 1
```
Comment (lucas-clemente) [inline ackhandler/received_packet_handler.go]: This looks weird – if we have a function SetLowerLimit and a var lowerLimit I'd expect them to deal with the same value. Can we remove this here and change the comparison above?

File: ackhandler/received_packet_handler.go [hunk 3]
```diff
@@ -39,31 +39,31 @@ func (h *receivedPacketHandler) ReceivedPacket(packetNumber protocol.PacketNumbe
...
+	}
 	h.maybeQueueAck(packetNumber, shouldInstigateAck)
 	return nil
 }
 
-func (h *receivedPacketHandler) ReceivedStopWaiting(f *frames.StopWaitingFrame) error {
-	// ignore if StopWaiting is unneeded, because we already received a StopWaiting with a higher LeastUnacked
-	if h.ignorePacketsBelow >= f.LeastUnacked {
-		return nil
+// SetLowerLimit sets a lower limit for acking packets
```
Comment (lucas-clemente) [inline ackhandler/received_packet_handler.go]: Please add punctuation and proper capitalization for godoc.

File: ackhandler/received_packet_history.go [hunk 4]
```diff
@@ -7,6 +7,8 @@ import (
 	"github.com/lucas-clemente/quic-go/qerr"
 )
 
+// the receivedPacketHistory stores if a packet number has already been received
```
Comment (lucas-clemente) [inline ackhandler/received_packet_history.go]: Please add punctuation and proper capitalization for godoc.

**Comment by marten-seemann** [comment]
I had hoped we could get some information about NSTP from the Google QUIC team, e.g. if they're still planning to implement that as a handshake tag, and if so, for which QUIC version.

File: ackhandler/received_packet_handler.go (hunk 1 above)
Comment (marten-seemann) [inline ackhandler/received_packet_handler.go]: That's true, but I think that's the correct behavior anyway. Acking non-retransmittable packets is going away with QUIC 39 by the way, so this is only temporary.

Comment (lucas-clemente) [inline ackhandler/received_packet_handler.go]: sgtm

**Comment by marten-seemann** [comment]
I asked on the proto-quic list: https://groups.google.com/a/chromium.org/forum/#!topic/proto-quic/nzDAV-pWuRU

**Comment by marten-seemann** [comment]
@lucas-clemente: I made the changes to achieve the more consistent behaviour of `receivedPacketHandler.SetLowerLimit`, and I added the guard against the NSTP CHLO tag.
Please have another look.

File: handshake/crypto_setup_server.go [hunk 5]
```diff
@@ -55,10 +55,14 @@ type cryptoSetupServer struct {
 var _ CryptoSetup = &cryptoSetupServer{}
 
 // ErrHOLExperiment is returned when the client sends the FHL2 tag in the CHLO
-// this is an expiremnt implemented by Chrome in QUIC 36, which we don't support
+// this is an experiment implemented by Chrome in QUIC 36, which we don't support
```
Comment (lucas-clemente) [inline handshake/crypto_setup_server.go]: Formatting ;)

File: handshake/crypto_setup_server.go [hunk 6]
```diff
@@ -55,10 +55,14 @@ type cryptoSetupServer struct {
...
 // TODO: remove this when dropping support for QUIC 36
 var ErrHOLExperiment = qerr.Error(qerr.InvalidCryptoMessageParameter, "HOL experiment. Unsupported")
 
+// ErrNSTPExperiment is returned when the client sends the NSTP tag in the CHLO
```
Comment (lucas-clemente) [inline handshake/crypto_setup_server.go]: Formatting ;)

File: ackhandler/received_packet_handler.go [hunk 7]
```diff
@@ -39,31 +39,29 @@ func (h *receivedPacketHandler) ReceivedPacket(packetNumber protocol.PacketNumbe
...
 
-func (h *receivedPacketHandler) ReceivedStopWaiting(f *frames.StopWaitingFrame) error {
-	// ignore if StopWaiting is unneeded, because we already received a StopWaiting with a higher LeastUnacked
-	if h.ignorePacketsBelow >= f.LeastUnacked {
-		return nil
+// SetLowerLimit sets a lower limit for acking packets.
+// Packets with packet numbers smaller or equal than p will not be acked.
+func (h *receivedPacketHandler) SetLowerLimit(p protocol.PacketNumber) {
+	h.lowerLimit = p
+	if p < protocol.MaxPacketNumber {
```
Comment (lucas-clemente) [inline ackhandler/received_packet_handler.go]: Can you please explain this / add a comment?

**Comment by marten-seemann** [comment]
@lucas-clemente: Fixed everything.

**Pull request merged**